Add variable scope to symbolic operations using TensorFlow

gonzaiva opened the pull request:
A variable scope facilitates the reading of a TensorFlow graph by grouping tensor objects in hierarchies.
In rllab, the hierarchies are defined with primitive objects and symbolic operations, where the latter is a member function of a
primitive.
In this change, the primitives are the algorithms, networks, distributions, optimizers, policies, Q-functions and regressors.
An example of a primitive is the DiagonalGaussian class, which
implements a probability distribution, and the symbolic function
is kl_sym, which implements the Kullback–Leibler divergence. The idea of implementing the variable scope is that all the tensor operations in kl_sym are encapsulated within the hierarchy DiagonalGaussian/kl_sym.
Each primitive and symbolic function have the parameter "name", which has a default value equal to the primitive or symbolic function name, but developers can set those parameters as they may find more convenient, so the previous example could be changed to be: distribution/divergence.
A context class was added to the file tensor_utils.py to verify if the variable scope of the corresponding primitive is already set, in order to set it in case it's not, and remove it once the symbolic function is exiting.
The only caveat with variable scopes is that they work based on the call stack and not with a class or file scope, so even if the same scope is used in two different call stacks, TensorFlow adds an index to the primitive or symbolic function names to make each scope unique. Therefore, not all symbolic operations performed in a primitive will appear under the same primitive scope, but they will be split in different instances of the primitive scope (e.g. DiagonalGaussian and
DiagonalGaussian_1).

Commit:
Add variable scope to symbolic operations using TensorFlow

A variable scope facilitates the reading of a TensorFlow graph
by grouping tensor objects in hierarchies.
In rllab, the hierarchies are defined with primitive objects and
symbolic operations, where the latter is a member function of a
primitive.
In this change, the primitives are the algorithms, networks,
distributions, optimizers, policies, Q-functions and regressors.
An example of a primitive is the DiagonalGaussian class, which
implements a probability distribution, and the symbolic function
is kl_sym, which implements the Kullback–Leibler divergence. The
idea of implementing the variable scope is that all the tensor
operations in kl_sym are encapsulated within the hierarchy
DiagonalGaussian/kl_sym.
Each primitive and symbolic function have the parameter "name",
which has a default value equal to the primitive or symbolic
function name, but developers can set those parameters as they
may find more convenient, so the previous example could be
changed to be: distribution/divergence.
A context class was added to the file tensor_utils.py to verify
if the variable scope of the corresponding primitive is already
set, in order to set it in case it's not, and remove it once the
symbolic function is exiting.
The only caveat with variable scopes is that they work based on
the call stack and not with a class or file scope, so even if
the same scope is used in two different call stacks, TensorFlow
adds an index to the primitive or symbolic function names
to make each scope unique. Therefore, not all symbolic
operations performed in a primitive will appear under the same
primitive scope, but they will be split in different instances
of the primitive scope (e.g. DiagonalGaussian and
DiagonalGaussian_1).

ghost commented:
To review this change, please uncompress the following tensorboard log:
[tensorboard_log.tar.gz](https://github.com/ryanjulian/rllab/files/2026130/tensorboard_log.tar.gz)
Then run the following command:
$ tensorboard --logdir="<log_path>"
Where log_path is the directory where you uncompressed the tar.gz file, and open the web address returned by tensorboard.

ghost commented:
Both build failures are due to the YAPF check:
![image](https://user-images.githubusercontent.com/35857569/40357320-a9f1b498-5d70-11e8-858e-2271d4167f2c.png)
However, no PEP8 formatting was introduced in this change to follow instructions given in #66.

ghost commented:
Sorry for being verbose, but the idea of executing tensorboard is to review the node graph along with the change to have a better understanding of the outcome with the proposed changed. The graph should look like:
![image](https://user-images.githubusercontent.com/35857569/40357877-5c3be424-5d72-11e8-9974-dfb80b829005.png)

ryanjulian commented:
Wow this looks great. Where does the top-level `Hx_plain_gradients` come from? Shouldn't it also end up inside `NPO`?

ryanjulian commented on sandbox/rocky/tf/misc/tensor_utils.py:
```diff
@@ -118,3 +118,21 @@ def pad_tensor_dict(tensor_dict, max_len):
             ret[k] = pad_tensor(tensor_dict[k], max_len)
     return ret
+
+class sys_op_scope(object):
```
`sys_op_scope()` is pretty generic and might be confused with a native TensorFlow feature. Maybe we can call it `enclosing_scope()`.

ryanjulian commented on sandbox/rocky/tf/misc/tensor_utils.py:
```diff
@@ -118,3 +118,21 @@ def pad_tensor_dict(tensor_dict, max_len):
         else:
             ret[k] = pad_tensor(tensor_dict[k], max_len)
     return ret
+
```
Please apply YAPF to this file, since you made substantial changes to it.

ryanjulian reviewed (commented):
This is looking great. The graphs generated are very clean.

For approval:
* See inline comments
* Can you explain the dangling `Hx_plain_gradients` node? Where does that come from? Should it have an enclosing scope?

Commit:
Rename sys_op_scope to enclosing_scope and format tensor_utils

The function sys_op_scope was renamed to enclosing_scope and the file
tensor_utils.py was formatted to PEP8.

gonzaiva commented on sandbox/rocky/tf/optimizers/conjugate_gradient_optimizer.py:
```diff
@@ -7,47 +7,51 @@
+            xs = tuple([tensor_utils.new_tensor_like(p.name.split(":")[0], p) for p in params])
+
+            def Hx_plain():
+                Hx_plain_splits = tf.gradients(
```
Although the tensor for Hx_plain_gradients is defined within the class scope of PerlmutterHvp, function Hx_plain() does not get call within the class, and therefore the variable scope of the primitive does not apply to Hx_plain_gradients

gonzaiva commented on sandbox/rocky/tf/optimizers/conjugate_gradient_optimizer.py:
```diff
@@ -7,47 +7,51 @@
+                return tensor_utils.flatten_tensor_variables(Hx_plain_splits)
+
+            self.opt_fun = ext.lazydict(
+                f_Hx_plain=lambda: tensor_utils.compile_function(
```
Function Hx_plain() is only passed as an anonymous function within a dictionary

gonzaiva replied:
Function Hx_plain() is only passed as an anonymous function within a dictionary

gonzaiva commented on sandbox/rocky/tf/optimizers/conjugate_gradient_optimizer.py:
```diff
@@ -7,47 +7,51 @@
+            xs = tuple([tensor_utils.new_tensor_like(p.name.split(":")[0], p) for p in params])
+
+            def Hx_plain():
+                Hx_plain_splits = tf.gradients(
```
Although the tensor for Hx_plain_gradients is defined within the class scope of PerlmutterHvp, function Hx_plain() does not get call within the class, and therefore the variable scope of the primitive does not apply to Hx_plain_gradients

ryanjulian replied:
Hmm this just seems to me like a special case with clear correct behavior. We can use the closure property to capture the correct enclosing scope during function definition. Then it doesn't matter whether it's called in a lambda or not.
```python
def Hx_plain():
    with sys_op_scope(name, "Hx_plain")
        Hx_plain_splits = tf.gradients( #...
```
And then even when it's invoked via lambda, self._name will have been captured during the definition and it will appear inside scope PerlmutterHvp.

ryanjulian replied:
Oh, I guess context managers don't cross function boundaries. That makes sense.

what if we did
```python
def Hx_plain():
    with sys_op_scope(self._name, name):
        with sys_op_scope(name, "Hx_plain"):
            Hx_plain_splits = tf.gradients( #...
```

I think that would make the expected output.

ghost commented:
The anonymous function is finally called in the function to calculate conjugate gradients, cg() in file rllab/misc/krylov.py.
`
def cg(f_Ax, b, cg_iters=10, callback=None, verbose=False, residual_tol=1e-10):
`
If we add a variable scope in that function, voilà!
![image](https://user-images.githubusercontent.com/35857569/40457978-46926a5c-5eae-11e8-8640-d0813fd0f479.png)
The scope expanded:
![image](https://user-images.githubusercontent.com/35857569/40457998-5b5bdca2-5eae-11e8-85ee-24b11debee19.png)
The only problem is that the cg function is in the Theano tree, so a TensorFlow import will break the isolation of both packages.

Commit:
Add a variable scope for the tensor Hx_plain_gradients

The tensor Hx_plain_gradients is defined in PerlmutterHvp but is not
created there. It's called in another file using a lamda function, so
the enclosing_scope function won't fully cover that scenario.
That's why the complete scope is passed as a parameter in the function
where Hx_plain_gradients is defined, so the scope can be reestablished
once the gradients are created.

ghost commented:
Reviewers,
Please review the new graph by uncompressing the following **updated** tensorboard log files:
[tensorflow_log.tar.gz](https://github.com/ryanjulian/rllab/files/2033631/tensorflow_log.tar.gz)
Then run the following command:
$ tensorboard --logdir="<log_path>"
Where log_path is the directory where you uncompressed the tar.gz file, and open the web address returned by tensorboard.

ryanjulian commented on sandbox/rocky/tf/misc/tensor_utils.py:
```diff
@@ -118,3 +120,25 @@ def pad_tensor_dict(tensor_dict, max_len):
+
+
+class enclosing_scope(object):
+    def __init__(self, primitive_name, sys_op_name, **kwargs):
```
primitive_name --> enclosing_name
sys_op_name --> name

...and similarly for the internal variables.

ryanjulian reviewed (approved):
Please address my last comment (naming) then merge.

Commit:
Rename members and parameters in class enclosing_scope

The pimitive_name and primitive_scope identifiers were changed to
enclosing_name and enclosing_scope respectively, while sys_op_name and
sys_op_scope were changed to simply name and scope.

ghost commented:
@ryanjulian, I did the change you requested, but I cannot merge due to the checks. Could you please help me with the merge? Thanks in advance.

Commit:
Add imports in GaussianLSTMPolicy

There is a missing import in
sandbox/rocky/tf/policies/gaussian_lstm_policy.py.
Add it.

ghost commented:
I verified the difference between the rebased commits and my merged-contaminated commits, and the only differences are the upgrade of Theano, the missing import that @CatherineSue found, thank you Chang, and the proper PEP8 formatting of the code I touched in this [commit](https://github.com/ryanjulian/rllab/pull/72/commits/9f9afd347c788dbbcb59162cff78c058ca1992e6).
I also tested the code and loaded the graph in tensorboard graph correctly from the new rebased commits. @ryanjulian, if there's something else to verify that I'm ignoring, please tell me, otherwise it's good to go.
Sorry about the merges and the rebase process. Should we consider creating a coding standard for developers in rllab to include standards and practices that we maybe ignoring @ryanjulian?
